File: Middleware/Thrift-homework/src/main/py/thrift_server.py
import os
import sys
import time
from thrift.transport import TSocket
from thrift.transport import TTransport
from thrift.protocol import TBinaryProtocol
from thrift.protocol import TCompactProtocol
from thrift.protocol import TJSONProtocol
from thrift.server import TServer

from src.main.py.gen_py.transfer import Transfer


class TransferHandler:
    """A class responsible for handling data transfers."""

    def __init__(self):
        self.log = {}

    def transferList(self, data):
        print(f"Received a list of {len(data)} elements")
        return int(round(time.time() * 1000))

    def transferSet(self, data):
        print(f"Received a set of {len(data)} elements")
        return int(round(time.time() * 1000))

    def transferMap(self, data):
        print(f"Received a map of {len(data)} elements")
        return int(round(time.time() * 1000))


def print_error():
    """A function printing an error when bad arguments provided"""

    print("Bad arguments provided.")
    print("The server needs a serialisation method to be provided as program argument: binary | json | compact")


def read_protocol(protocol_type):
    """
    A function returning the protocol factory based on an argument provided.

    :param protocol_type: the protocol type
    :return: the protocol factory of the required type
    """

    if protocol_type == "binary":
        return TBinaryProtocol.TBinaryProtocolFactory()
    elif protocol_type == "json":
        return TJSONProtocol.TJSONProtocolFactory()
    elif protocol_type == "compact":
        return TCompactProtocol.TCompactProtocolFactory()
    return None


if __name__ == '__main__':
    """The simple Thrift server receiving data from the Thrift client."""

    if len(sys.argv) < 2:
        print_error()
        exit(1)

    handler = TransferHandler()
    processor = Transfer.Processor(handler)
    transport = TSocket.TServerSocket(host='127.0.0.1', port=9080)
    transport_factory = TTransport.TBufferedTransportFactory()

    protocol_type = sys.argv[1]
    protocol_factory = read_protocol(protocol_type)
    if protocol_factory is None:
        print_error()
        exit(1)

    server = TServer.TSimpleServer(processor, transport, transport_factory, protocol_factory)

    print('The server is started...')

    try:
        server.serve()
    except KeyboardInterrupt:
        print('A server is interrupted')
        try:
            sys.exit(0)
        except SystemExit:
            os._exit(0)

File: Middleware/Thrift-homework/src/main/java/xenoteo/com/github/server/ThriftServer.java
package xenoteo.com.github.server;

import org.apache.thrift.protocol.TBinaryProtocol;
import org.apache.thrift.protocol.TCompactProtocol;
import org.apache.thrift.protocol.TJSONProtocol;
import org.apache.thrift.protocol.TProtocolFactory;
import org.apache.thrift.server.TServer;
import org.apache.thrift.server.TSimpleServer;
import org.apache.thrift.transport.TServerSocket;
import org.apache.thrift.transport.TServerTransport;
import xenoteo.com.github.thrift.gen.Transfer;

/**
 * The simple Thrift server receiving data from the Thrift client.
 */
public class ThriftServer {

    public static void main(String[] args) {
        try {
            if (args.length < 1) {
                printError();
                System.exit(1);
            }

            TProtocolFactory protocolFactory = readProtocol(args[0]);
            if (protocolFactory == null) {
                printError();
                System.exit(1);
            }

            Transfer.Processor<TransferHandler> processor = new Transfer.Processor<>(new TransferHandler());

            TServerTransport serverTransport = new TServerSocket(9080);

            TServer server = new TSimpleServer(new TServer.Args(serverTransport).protocolFactory(protocolFactory).processor(processor));

            System.out.println("Starting the simple server...");
            server.serve();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    /**
     * Prints an error when bad arguments provided.
     */
    private static void printError() {
        System.err.println("Bad arguments provided.");
        System.err.println("The server needs a serialisation method to be provided as program argument: " +
                "binary | json | compact");
    }

    /**
     * Returns the TProtocolFactory based on an argument provided.
     *
     * @param protocol  the string representing a protocol type
     * @return the TProtocolFactory of the required type
     */
    private static TProtocolFactory readProtocol(String protocol){
        return switch (protocol) {
            case "binary" -> new TBinaryProtocol.Factory();
            case "json" -> new TJSONProtocol.Factory();
            case "compact" -> new TCompactProtocol.Factory();
            default -> null;
        };
    }

}
